src/http.nim

```diff
@@ -0,0 +1,250 @@
+      return Error
+    else:
+      return ErrorFailure
+
+  return Success
+
+proc serveClient(server: StreamServer, transp: StreamTransport) {.async.} =
+  ## Process transport data to the RPC server
+  var httpServer = cast[HttpServer](server)
+  var buffer = newSeq[byte](MaxHttpHeadersSize)
+  var header: HttpRequestHeader
+
+  info "Received connection", address = $transp.remoteAddress()
+  try:
+    let hlenfut = transp.readUntil(addr buffer[0], MaxHttpHeadersSize,
+        sep = HeaderSep)
+    let ores = await withTimeout(hlenfut, HttpHeadersTimeout)
+    if not ores:
+      # Timeout
+      debug "Timeout expired while receiving headers",
+            address = transp.remoteAddress()
+      discard await transp.sendHTTPResponse(HttpVersion11, Http408)
+      await transp.closeWait()
+      return
+    else:
+      let hlen = hlenfut.read()
+      buffer.setLen(hlen)
+      header = buffer.parseRequest()
+      if header.failed():
+        # Header could not be parsed
+        debug "Malformed header received",
+              address = transp.remoteAddress()
+        discard await transp.sendHTTPResponse(HttpVersion11, Http400)
+        await transp.closeWait()
+        return
+      var vres = await validateRequest(transp, header)
+      if vres == Success:
+        info "Received valid RPC request", address = $transp.remoteAddress()
+        # Call the user's callback.
+        if httpServer.callback != nil:
+          await httpServer.callback(transp, header)
+      elif vres == ErrorFailure:
+        debug "Remote peer disconnected", address = transp.remoteAddress()
+  except TransportLimitError:
+    # size of headers exceeds `MaxHttpHeadersSize`
+    debug "Maximum size of headers limit reached",
+          address = transp.remoteAddress()
+    discard await transp.sendHTTPResponse(HttpVersion11, Http413)
+  except TransportIncompleteError:
+    # remote peer disconnected
+    debug "Remote peer disconnected", address = transp.remoteAddress()
+  except TransportOsError as exc:
+    debug "Problems with networking", address = transp.remoteAddress(),
+          error = exc.msg
+  except CatchableError as exc:
+    debug "Unknown exception", address = transp.remoteAddress(),
+          error = exc.msg
+  await transp.closeWait()
+
+proc newHttpServer*(address: string, handler: AsyncCallback,
+              flags: set[ServerFlags] = {ReuseAddr}): HttpServer =
+  let address = initTAddress(address)
+  var server = HttpServer(callback: handler)
+  server = cast[HttpServer](createStreamServer(address, serveClient, flags,
+      child = cast[StreamServer](server)))
+  return server
+
+func toTitleCase(s: string): string =
+  var tcstr = newString(len(s))
+  var upper = true
+  for i in 0..len(s) - 1:
+    tcstr[i] = if upper: toUpperAscii(s[i]) else: toLowerAscii(s[i])
+    upper = s[i] == '-'
+  return tcstr
+
+func toCaseInsensitive*(headers: HttpHeaders, s: string): string {.inline.} =
+  return toTitleCase(s)
```

Comment:
Maybe just toLower/toUpper case?

Reply:
HTTP header is not case sensitive. Hence converting it to title case for better readability.

Reply:
Yeah, I meant that `toTitleCase` is replaced with `toUpper` or `toLower` since case is irrelevant here. `toTitleCase` seems to be doing too much by allocating a buffer and copying data just for the sake of capitalizing the first letter?